File: Sudoku-go/src/SudokuByHand.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sync"
	"time"
	"unicode"
	dafnySolver "SudokuSolver"
	"dafny"
)

const (
	gridSize   = 9
	totalCells = gridSize * gridSize
)

type Board [gridSize][gridSize]byte

func loadPuzzlesFromFile(filePath string) ([]Board, []Board, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, nil, fmt.Errorf("error: the file '%s' was not found: %w", filePath, err)
	}
	defer file.Close()

	boards1Chan := make(chan Board)
	boards2Chan := make(chan Board)
	var wg sync.WaitGroup

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		wg.Add(1)
		go func(l string) {
			defer wg.Done()
			if board, ok := tryParseBoard(l); ok {
				boards1Chan <- board
				boards2Chan <- board
			}
		}(line)
	}

	go func() {
		wg.Wait()
		close(boards1Chan)
		close(boards2Chan)
	}()

	var boards1, boards2 []Board
	var collectionWg sync.WaitGroup
	collectionWg.Add(2)
	go func() {
		defer collectionWg.Done()
		for board := range boards1Chan {
			boards1 = append(boards1, board)
		}
	}()
	go func() {
		defer collectionWg.Done()
		for board := range boards2Chan {
			boards2 = append(boards2, board)
		}
	}()
	collectionWg.Wait()


	if err := scanner.Err(); err != nil {
		return nil, nil, fmt.Errorf("error reading file: %w", err)
	}

	return boards1, boards2, nil
}

func tryParseBoard(puzzleString string) (Board, bool) {
	if len(puzzleString) != totalCells {
		return Board{}, false
	}

	var board Board
	for i := 0; i < totalCells; i++ {
		cellChar := rune(puzzleString[i])
		if !unicode.IsDigit(cellChar) {
			return Board{}, false
		}
		row := i / gridSize
		col := i % gridSize
		// Conversion from char to byte digit.
		board[row][col] = byte(cellChar - '0')
	}
	return board, true
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run main.go <puzzle_file>")
		return
	}
	puzzleFile := os.Args[1]

	boards1, boards2, err := loadPuzzlesFromFile(puzzleFile)
	if err != nil {
		log.Fatalf("Failed to load puzzles: %v", err)
	}
	
	if boards1 == nil || boards2 == nil {
        fmt.Println("Something went wrong when reading the file, results are null")
        return
    }

	test(boards1, boards2)
}

func test(boards1, boards2 []Board) {
	start1 := time.Now()
	run(boards1)
	duration1 := time.Since(start1)
	fmt.Printf("go_handwritten,%d,%d\n", len(boards1), duration1.Milliseconds())

	start2 := time.Now()
	dafnySolver.Companion_Default___.Run(convertToGoDafnyArrays(boards2))
	duration2 := time.Since(start2)
	fmt.Printf("go_dafny,%d,%d\n", len(boards2), duration2.Milliseconds())
}

func convertToGoDafnyArrays(boards []Board) dafny.Array {
	dafnyBoards := dafny.NewArray(dafny.IntOf(len(boards)))

	dim := dafny.IntOf(gridSize)

	for i, goBoard := range boards {
		dafnyBoard := dafny.NewArrayFromExample(byte(0), nil, dim, dim)

		for r := 0; r < gridSize; r++ {
			for c := 0; c < gridSize; c++ {
				flatIndex := r*gridSize + c
				value := goBoard[r][c]

				dafnyBoard.ArraySet1Byte(value, flatIndex)
			}
		}
		dafnyBoards.ArraySet1(dafnyBoard, i)
	}

	return dafnyBoards
}

func run(boards []Board) []bool {
	results := make([]bool, len(boards))
	for i := range boards {
		_, ok := solve(&boards[i])
		results[i] = ok
	}
	return results
}

func solve(board *Board) (*Board, bool) {
	if !isValidBoard(board) {
		return nil, false
	}
	return solving(board)
}

func solving(board *Board) (*Board, bool) {
	r, c, ok := findEmptySlot(board)
	if !ok {
		return board, true
	}

	for digit := byte(1); digit <= 9; digit++ {
		if isValidDigit(board, r, c, digit) {
			board[r][c] = digit
			if solvedBoard, ok := solving(board); ok {
				return solvedBoard, true
			}
			board[r][c] = 0
		}
	}

	return nil, false
}

func findEmptySlot(board *Board) (row, col byte, ok bool) {
	for r := byte(0); r < gridSize; r++ {
		for c := byte(0); c < gridSize; c++ {
			if board[r][c] == 0 {
				return r, c, true
			}
		}
	}
	return 0, 0, false
}

func isValidBoard(board *Board) bool {
	for r := byte(0); r < gridSize; r++ {
		for c := byte(0); c < gridSize; c++ {
			if !isValidDigit(board, r, c, board[r][c]) {
				return false
			}
		}
	}
	return true
}

func isValidDigit(board *Board, row, col, digit byte) bool {
	return isValidInRowAndColumn(board, row, col, digit) && isValidIn3x3(board, row, col, digit)
}

func isValidInRowAndColumn(board *Board, row, col, digit byte) bool {
	if digit == 0 {
		return true
	}
	for b := byte(0); b < gridSize; b++ {
		if b != col && board[row][b] == digit {
			return false
		}
		if b != row && board[b][col] == digit {
			return false
		}
	}
	return true
}

func isValidIn3x3(board *Board, row, col, digit byte) bool {
	if digit == 0 {
		return true
	}
	boxRowStart := row / 3 * 3
	boxColStart := col / 3 * 3

	for r := boxRowStart; r < boxRowStart+3; r++ {
		for c := boxColStart; c < boxColStart+3; c++ {
			if r == row && c == col {
				continue
			}
			if board[r][c] == digit {
				return false
			}
		}
	}
	return true
}

File: Sudoku-py/SudokuByHand.py
import sys
import os
import time
import copy
from enum import Enum
from typing import Optional, List, Tuple
from concurrent.futures import ProcessPoolExecutor
from SudokuSolver import default__
import _dafny

Board = List[List[int]]

GRID_SIZE = 9
TOTAL_CELLS = GRID_SIZE * GRID_SIZE

def try_parse_board(puzzle_string: str) -> Optional[Board]:
    if not puzzle_string or len(puzzle_string) != TOTAL_CELLS:
        return None

    board = [[0] * GRID_SIZE for _ in range(GRID_SIZE)]
    for i in range(TOTAL_CELLS):
        cell_char = puzzle_string[i]
        if not cell_char.isdigit():
            return None
        
        row = i // GRID_SIZE
        col = i % GRID_SIZE
        board[row][col] = int(cell_char)
        
    return board

def load_puzzles_from_file(file_path: str) -> Tuple[List[Board], List[Board]]:
    if not os.path.exists(file_path):
        raise FileNotFoundError(f"Error: The file '{file_path}' was not found.")

    boards1 = []
    
    with open(file_path, 'r') as f, ProcessPoolExecutor() as executor:
        lines_generator = (line.strip() for line in f if line.strip())
        
        results = executor.map(try_parse_board, lines_generator, chunksize=100)
        
        boards1 = [board for board in results if board is not None]

    boards2 = copy.deepcopy(boards1)
    
    return boards1, boards2


class Option:
    def __init__(self, value, is_some: bool):
        self._value = value
        self.is_some = is_some

    @property
    def value(self):
        if not self.is_some:
            raise ValueError("Cannot access the value of a None option.")
        return self._value

    @staticmethod
    def Some(value):
        return Option(value, True)

    @staticmethod
    def None_():
        return Option(None, False)

def is_9x9(board: Board) -> bool:
    return len(board) == 9 and all(len(row) == 9 for row in board)

def is_valid_in_row_and_column(board: Board, row: int, column: int, digit: int) -> bool:
    if digit == 0:
        return True
    for i in range(9):
        if i != column and board[row][i] == digit:
            return False
        if i != row and board[i][column] == digit:
            return False
    return True

def is_valid_in_3x3(board: Board, row: int, column: int, digit: int) -> bool:
    if digit == 0:
        return True
    box_row_start = (row // 3) * 3
    box_col_start = (column // 3) * 3
    
    for r in range(box_row_start, box_row_start + 3):
        for c in range(box_col_start, box_col_start + 3):
            if r == row and c == column:
                continue
            if board[r][c] == digit:
                return False
    return True

def is_valid_digit(board: Board, row: int, column: int, digit: int) -> bool:
    return is_valid_in_row_and_column(board, row, column, digit) and is_valid_in_3x3(board, row, column, digit)

def is_valid_board(board: Board) -> bool:
    for r in range(len(board)):
        for c in range(len(board[0])):
            digit = board[r][c]
            board[r][c] = 0
            if not is_valid_digit(board, r, c, digit):
                board[r][c] = digit 
                return False
            board[r][c] = digit
    return True

def find_empty_slot(board: Board) -> Option:
    for r in range(9):
        for c in range(9):
            if board[r][c] == 0:
                return Option.Some((r, c))
    return Option.None_()

def _solving(board: Board) -> Option:
    empty = find_empty_slot(board)
    if not empty.is_some:
        return Option.Some(board)
    
    r, c = empty.value
    
    for digit in range(1, 10):
        if is_valid_digit(board, r, c, digit):
            board[r][c] = digit
            
            recursive_result = _solving(board)
            if recursive_result.is_some:
                return recursive_result
            
            board[r][c] = 0
            
    return Option.None_()

def solve(board: Board) -> Option:
    if not is_9x9(board):
        return Option.None_()
    
    if not is_valid_board(board):
        return Option.None_()

    return _solving(board)

def run(boards: List[Board]) -> List[bool]:
    results = []
    for board in boards:
        result = solve(board)
        results.append(result.is_some)
    return results



def convert_to_dafny_array(python_boards: List[Board]) -> _dafny.Array:
    num_boards = len(python_boards)
    dafny_boards = _dafny.Array(None, num_boards, GRID_SIZE, GRID_SIZE)

    for i in range(num_boards):
        dafny_boards[i] = _dafny.Array(None, GRID_SIZE, GRID_SIZE)
        for j in range(GRID_SIZE):
            for k in range(GRID_SIZE):
                dafny_boards[i, j, k] = python_boards[i][j][k]
    
    return dafny_boards

def test(boards1: List[Board], boards2: List[Board]):
    """ Times the execution of the solver. """
    sw1_start = time.perf_counter()
    run(boards1)
    sw1_stop = time.perf_counter()
    elapsed_ms1 = (sw1_stop - sw1_start) * 1000
    print(f"python_handwritten,{len(boards1)},{int(elapsed_ms1)}")

    sw2_start = time.perf_counter()
    default__.Run(convert_to_dafny_array(boards2))
    sw2_stop = time.perf_counter()
    elapsed_ms2 = (sw2_stop - sw2_start) * 1000
    print(f"python_dafny,{len(boards2)},{int(elapsed_ms2)}")


def main():
    if len(sys.argv) != 2:
        print("Usage: python your_script_name.py <puzzle_file>")
        return

    puzzle_file = sys.argv[1]
    
    boards1, boards2 = load_puzzles_from_file(puzzle_file)
    if boards1 is None or boards2 is None:
        print("Something went wrong when reading the file, results are null")
        return
    test(boards1, boards2)

if __name__ == "__main__":
    main()